goto_handler.go

```diff
@@ -0,0 +1,116 @@
+package migrate
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+func (m *Migrate) HandleDirtyState(path string) {
+	// Perform actions when dirty is true
+		lastSuccessfulMigrationPath := filepath.Join(path, "lastSuccessfulmigration")
+		lastVersionBytes, err := os.ReadFile(lastSuccessfulMigrationPath)
+		if err != nil {
+			log.Fatal(fmt.Errorf("failed to read last successful migration file: %s", err))
+		}
+		lastVersionStr := strings.TrimSpace(string(lastVersionBytes))
+		lastVersion, err := strconv.ParseUint(lastVersionStr, 10, 64)
+		if err != nil {
+			log.Fatal(fmt.Errorf("failed to parse last successful migration version: %s", err))
+		}
+
+		if err := m.Force(int(lastVersion)) ; err != nil {
+			log.Fatal(fmt.Errorf("failed to apply last successful migration: %s", err))
+		}
+
+		fmt.Println("Last successful migration applied")
+
+		if err := os.Remove(lastSuccessfulMigrationPath); err != nil {
+			log.Fatal(fmt.Errorf("failed to delete last successful migration file: %s", err))
+		}
+
+		fmt.Println("Last successful migration file deleted")
+
+}
+
+func (m *Migrate) HandleMigrationFailure(curVersion int, v uint, path string) ([]int, error) {
+	if err := m.lock(); err != nil {
```

Comment:
Mutex has to be unlocked before existing the function. See the `unlockErr()` function.

Reply:
The `unlockErr()` function has to be called every time an error is returned. Use an existing function using it as an example you can follow.